internal/dump/acl.go

```diff
@@ -1,131 +1,88 @@
-		str += "-"
-	}
-	return fmt.Sprintf("%v%v", a.aclSID, str)
-}
+	// Tags.
+	aclTagUserObj  = 0x01 // Owner.
+	aclTagUser     = 0x02
+	aclTagGroupObj = 0x04 // Owning group.
+	aclTagGroup    = 0x08
+	aclTagMask     = 0x10
+	aclTagOther    = 0x20
+)
 
-func (a *acl) decode(xattr []byte) {
-	var elem aclElement
-	ae := new(aclElem)
-	nr := bytes.NewReader(xattr)
-	e := binary.Read(nr, binary.LittleEndian, &a.Version)
-	if e != nil {
-		a.Version = 0
-		return
-	}
-	if len(a.List) > 0 {
-		a.List = a.List[:0]
+// formatLinuxACL converts a Linux ACL from its binary format to the POSIX.1e
+// long text format.
+//
+// User and group IDs are printed in decimal, because we may be dumping
+// a snapshot from a different machine.
+//
+// https://man7.org/linux/man-pages/man5/acl.5.html
+// https://savannah.nongnu.org/projects/acl
+// https://simson.net/ref/1997/posix_1003.1e-990310.pdf
+func formatLinuxACL(acl []byte) (string, error) {
+	if len(acl)-4 < 0 || (len(acl)-4)%8 != 0 {
+		return "", errors.New("wrong length")
 	}
-	for binary.Read(nr, binary.LittleEndian, ae) == nil {
-		elem.aclSID = (aclSID(ae.Tag) << 32) | aclSID(ae.ID)
-		elem.Perm = ae.Perm
-		a.List = append(a.List, elem)
+	version := binary.LittleEndian.Uint32(acl)
+	if version != 2 {
+		return "", errors.New("wrong version")
 	}
-}
-
-func (a *acl) encode() []byte {
-	buf := new(bytes.Buffer)
-	ae := new(aclElem)
+	acl = acl[4:]
 
-	err := binary.Write(buf, binary.LittleEndian, &a.Version)
-	// write to a bytes.Buffer always returns a nil error
-	if err != nil {
-		panic(err)
-	}
+	text := make([]byte, 0, 2*len(acl))
 
-	for _, elem := range a.List {
-		ae.Tag = uint16(elem.getType())
-		ae.Perm = elem.Perm
-		ae.ID = elem.getID()
+	for ; len(acl) >= 8; acl = acl[8:] {
+		tag := binary.LittleEndian.Uint16(acl)
+		perm := binary.LittleEndian.Uint16(acl[2:])
+		id := binary.LittleEndian.Uint32(acl[4:])
 
-		err := binary.Write(buf, binary.LittleEndian, ae)
-		// write to a bytes.Buffer always returns a nil error
-		if err != nil {
-			panic(err)
+		switch tag {
+		case aclTagUserObj:
+			text = append(text, "user:"...)
+		case aclTagUser:
+			text = append(text, "user:"...)
+			text = strconv.AppendUint(text, uint64(id), 10)
+		case aclTagGroupObj:
+			text = append(text, "group:"...)
+		case aclTagGroup:
+			text = append(text, "group:"...)
+			text = strconv.AppendUint(text, uint64(id), 10)
+		case aclTagMask:
+			text = append(text, "mask:"...)
+		case aclTagOther:
+			text = append(text, "other:"...)
+		default:
+			return "", errors.New("unknown tag")
 		}
+		text = append(text, ':')
+		text = append(text, aclPermText(perm)...)
+		text = append(text, '\n')
 	}
-	return buf.Bytes()
+
+	return string(text), nil
 }
 
-func (a *acl) String() string {
-	var finalacl string
-	for _, acl := range a.List {
-		finalacl += acl.String() + "\n"
+func aclPermText(p uint16) []byte {
+	s := []byte("---")
+	if p&aclPermRead != 0 {
+		s[0] = 'r'
+	}
+	if p&aclPermWrite != 0 {
+		s[0] = 'w'
```

Comment:
The offsets are wrong, for write it should be `1` and `2` for execute.

Reply:
Fixed, tests amended (and tested again against an actual ACL).